packages/x-date-pickers/src/DateField/DateField.tsx

```diff
@@ -51,7 +55,15 @@ const DateField = React.forwardRef(function DateField<TDate>(
     <TextField
       ref={ref}
       {...fieldProps}
-      InputProps={{ ...fieldProps.InputProps, readOnly }}
+      InputProps={{
+        ...fieldProps.InputProps,
+        readOnly,
+        endAdornment: clearable ? (
```

Comment:
We'd probably need a similar implementation to the one that Autocomplete uses, where the adornment is added when there is a value present.
But that wouldn't fix the existing issue of layout shift happening when the button is added.
Maybe if we would add the `endAdornment` always when a `clearable` prop is present, but only showing it then necessary would be the quickest solution... 🤔 
P.S. We should not forget that users might want to provide their own `endAdornment` and that needs to work in one way or another, currently it would be ignored.